### PR TITLE
doc: migrate lib.attrsets to use doc-comments

### PR DESCRIPTION
**Pull request by hsjobeki**
## Description of changes

**Doc-comment migration**

cc @DanielSidhion 

I am going to manually change the following after a the first round of review is done.

- [x] Inner/outer format migrated `/* {nixdoc} */` -> `/** {commonmark} */`
- [x] Clean up function arguments. Manually remove the redundant line comments. We decided the codemod won't delete anything.
- [x] Reify examples. After the migration is done all examples should be placed at the end under an `# Examples` heading.
    - [x] Examples should be surrunded by :::{.example} such that they are properly styled.
    - [x] Make sure examples have a descriptive heading. e.g. `## lib.asserts.assertMsg usage examples`
- [x] Manual reformatting work for consistency:
  - [x] foldlAttrs: Attention block needs proper highlighting
  - [x] Indentation: getAttrFromPath, 
  - [x] Whitespace: attrsToList
  - [x] mapAttrsRecursive / mapAttrsRecursiveCond: inconsistent order of example / type. (Wont fix)

Please comment all other errors / mistakes.

NOTE: PRs should not add/change content (excluding minor fixes) so we keep track of eventual rendering problems arising with the migration in isolation.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by hsjobeki** [comment]
@infinisil @DanielSidhion I think this PR should be ready to merge. If you'd give it a rough review.

I need to fix this in nixdoc: https://github.com/nix-community/nixdoc/issues/113 but shouldn't be a blocker.

File: lib/attrsets.nix [hunk 1]
```diff
@@ -342,96 +479,198 @@ rec {
...
-    # The set to get the named attributes from
     attrs: genAttrs names (name: attrs.${name});
 
-  /* Collect each attribute named `attr` from a list of attribute
-     sets.  Sets that don't contain the named attribute are ignored.
+  /**
+    Collect each attribute named `attr` from a list of attribute
+    sets.  Sets that don't contain the named attribute are ignored.
+
+    # Type
```
Comment (DanielSidhion) [inline lib/attrsets.nix:582]:
```suggestion
    # Inputs
    
    `attr`
    
    : The attribute name to get out of the sets.
    
    `list`
    
    : The list of attribute sets to go through
    
    # Type
```

File: lib/attrsets.nix [hunk 2]
```diff
@@ -579,76 +903,145 @@ rec {
...
 
-  /* Apply a function to each element in an attribute set, creating a new attribute set.
+  /**
+    Apply a function to each element in an attribute set, creating a new attribute set.
 
-     Example:
-       mapAttrs (name: value: name + "-" + value)
-          { x = "foo"; y = "bar"; }
-       => { x = "x-foo"; y = "y-bar"; }
+    # Type
```
Comment (DanielSidhion) [inline lib/attrsets.nix:946]:
```suggestion
    # Inputs
    
    `f`
    
    : A function that takes an attribute name and its value, and returns the new value for the attribute.
    
    `attrset`
    
    : The attribute set to iterate through.
    
    # Type
```

File: lib/attrsets.nix [hunk 3]
```diff
@@ -668,13 +1061,22 @@ rec {
     This is because the `listToAttrs` removes duplicate names and doesn't preserve the order of the list.
     :::
 
-    Example:
-      attrsToList { foo = 1; bar = "asdf"; }
-      => [ { name = "bar"; value = "asdf"; } { name = "foo"; value = 1; } ]
+    # Type
```
Comment (DanielSidhion) [inline lib/attrsets.nix:1064]:
```suggestion
    # Inputs
    
    `set`
    
    : The attribute set to deconstruct.
    
    # Type
```

File: lib/attrsets.nix [hunk 4]
```diff
@@ -1034,161 +1621,310 @@ rec {
...
       then pkg.${output} or pkg.out or pkg
       else pkg;
 
-  /* Get a package's `bin` output.
-     If the output does not exist, fallback to `.out` and then to the default.
+  /**
+    Get a package's `bin` output.
+    If the output does not exist, fallback to `.out` and then to the default.
+
+    # Type
```
Comment (DanielSidhion) [inline lib/attrsets.nix:1745]:
```suggestion
    # Inputs
    
    `pkg`
    
    : The package whose `bin` output will be retrieved.
    
    # Type
```

File: lib/attrsets.nix [hunk 5]
```diff
@@ -1034,161 +1621,310 @@ rec {
...
   getBin = getOutput "bin";
 
 
-  /* Get a package's `lib` output.
-     If the output does not exist, fallback to `.out` and then to the default.
+  /**
+    Get a package's `lib` output.
+    If the output does not exist, fallback to `.out` and then to the default.
+
+    # Type
```
Comment (DanielSidhion) [inline lib/attrsets.nix:1769]:
```suggestion
    # Inputs
    
    `pkg`
    
    : The package whose `lib` output will be retrieved.
    
    # Type
```

File: lib/attrsets.nix [hunk 6]
```diff
@@ -1034,161 +1621,310 @@ rec {
...
   getLib = getOutput "lib";
 
 
-  /* Get a package's `dev` output.
-     If the output does not exist, fallback to `.out` and then to the default.
+  /**
+    Get a package's `dev` output.
+    If the output does not exist, fallback to `.out` and then to the default.
+
+    # Type
```
Comment (DanielSidhion) [inline lib/attrsets.nix:1793]:
```suggestion
    # Inputs
    
    `pkg`
    
    : The package whose `dev` output will be retrieved.
    
    # Type
```

File: lib/attrsets.nix [hunk 7]
```diff
@@ -1034,161 +1621,310 @@ rec {
...
-  /* Get a package's `man` output.
-     If the output does not exist, fallback to `.out` and then to the default.
+  /**
+    Get a package's `man` output.
+    If the output does not exist, fallback to `.out` and then to the default.
 
-     Example:
-       getMan pkgs.openssl
-       => "/nix/store/9rz8gxhzf8sw4kf2j2f1grr49w8zx5vj-openssl-1.0.1r-man"
+    # Type
```
Comment (DanielSidhion) [inline lib/attrsets.nix:1817]:
```suggestion
    # Inputs
    
    `pkg`
    
    : The package whose `man` output will be retrieved.
    
    # Type
```

File: lib/attrsets.nix [hunk 8]
```diff
@@ -1034,161 +1621,310 @@ rec {
...
+    :::
   */
   getMan = getOutput "man";
 
-  /* Pick the outputs of packages to place in `buildInputs`
+  /**
+    Pick the outputs of packages to place in `buildInputs`
 
-   Type: chooseDevOutputs :: [Derivation] -> [String]
+    # Type
```
Comment (DanielSidhion) [inline lib/attrsets.nix:1839]:
```suggestion
    # Inputs
    
    `pkgs`
    
    : List of packages.
    
    # Type
```

**Review by DanielSidhion — APPROVED** [review]
Looks good to me! No information lost and the formatting is great.

While reading through the changes I made a few suggestions of content to add (since the effort to write the suggestions was small and we're changing this stuff already).

**Pull request merged**